ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol/Util.hs

```diff
@@ -28,3 +29,12 @@ isNewEpoch ei newSlot referenceWO = runIdentity $ do
     reference = fromWithOrigin genesisSlotNo referenceWO
     -- TODO
     genesisSlotNo = SlotNo 0
+
+-- | Return the first slot in the epoch of the given slot.
+firstSlotOfEpochOfSlot ::
```

Comment:
I anticipate that this could fit in a more general `Util` module, if you're interested in searching for other occurrences of this pattern. I'm sure I've written it multiple times, but I'm not sure those were merged/have survived.

Reply:
Could add it to `cardano-slotting`